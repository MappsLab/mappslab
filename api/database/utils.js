// @flow

import * as R from 'ramda'
import type { Filter, PaginationArgs } from 'Types/sharedTypes'

export const createVariables = R.pipe(
	R.toPairs,
	R.reduce(
		(acc, [key, value]) =>
			// if the value is undefined or null,
			value === undefined || value === null
				? // return the accumulator
				  acc
				: // Else, return the value cast to a string
				  R.assoc(`$${key}`, value.toString())(acc),
		{},
	),
)

export const defaultPaginationArgs = {
	first: 50,
	after: '0x0',
}

export const makePaginationArgs = (args: PaginationArgs): PaginationArgs => {
	// Adds +1 to the 'first' argument, we need this to see if there is a next page.
	const { first, ...rest } = { ...defaultPaginationArgs, ...args }
	return {
		first: first + 1,
		...rest,
	}
}

export const createFilterString = (filter: Array<Filter>, connect: string = 'AND'): string => {
	const funcs = filter.map(({ key, value, operator = 'eq' }) => `${operator}(${key}, "${value}")`)
	return `@filter(${funcs.join(connect)})`
}

export const validateUid = (uid: string): boolean => /^0x[\d\w]+$/.test(uid)

export const validateUidOrWildcard = (uid: string): boolean => uid === '*' || validateUid(uid)
