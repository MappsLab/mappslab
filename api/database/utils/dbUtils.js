// @flow

import * as R from 'ramda'
import type { Filter, PaginationFilterArgs } from 'Types/sharedTypes'
import makeFilterString from './makeFilterString'

/**
 * Properly formats incoming variables with '$' for dgraph
 */

export const createVariables = R.pipe(
	R.toPairs,
	R.reduce(
		(acc, [key, value]) =>
			// if the value is undefined or null,
			value === undefined || value === null
				? // return the accumulator
				  acc
				: // Else, return the value cast to a string
				  R.assoc(`$${key}`, value.toString())(acc),
		{},
	),
)

/**
 * Create a paginatino string to append to a dGraph func. Be sure to include the
 * initial comma that appears after the root function, i.e. `, first: 5, after: 0x123`
 * @param {number} first
 * @param {string} after
 */
export const DEFAULT_PAGE_COUNT = 50

const makePaginationString = (first?: number = DEFAULT_PAGE_COUNT, after?: string): string => {
	return [
		'', // include a blank string so we start it with a comma
		first // if a 'first' value is present, include it +1 so we can tell if there is a 'nextPage'
			? `first: ${first + 1}`
			: null,
		after //
			? `after: ${after}`
			: null,
	]
		.filter((i) => i !== null)
		.join(', ')
}

/**
 * Create a @filter string to append to a dGraph query. Be sure to include the
 * initial comma that appears after the root function, i.e. `, first: 5, after: 0x123`
 * @param {number} first
 * @param {string} after
 */

export type FilterStrings = {
	filterString: string,
	varBlocks: string,
}

type QueryStrings = FilterStrings & {
	paginationString: string,
}

export const createQueryStrings = (args: PaginationFilterArgs, filterDeleted?: boolean = true): QueryStrings => {
	if (!args)
		return {
			paginationString: '',
			filterString: '@filter((eq(deleted, false)))',
			varBlocks: '',
		}
	const { first, after, where: baseWhere } = args
	const where = filterDeleted
		? {
				deleted: { eq: false },
				...baseWhere,
		  }
		: baseWhere

	return {
		paginationString: makePaginationString(first, after),
		...makeFilterString(where),
	}
}

/**
 * Validate that the string is a dGraph uid
 */

export const validateUid = (uid: string): boolean => /^0x[\d\w]+$/.test(uid)

export const validateUidOrWildcard = (uid: string): boolean => uid === '*' || validateUid(uid)
