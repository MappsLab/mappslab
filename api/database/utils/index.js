export { default as makeFilterString } from './makeFilterString'
export * from './dbUtils'
