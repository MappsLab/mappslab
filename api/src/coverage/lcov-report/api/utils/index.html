<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for api/utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> api/utils
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">84.21% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>48/57</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">84.85% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>28/33</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">81.82% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>18/22</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">85.11% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>40/47</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="auth.js"><a href="auth.js.html">auth.js</a></td>
	<td data-value="85" class="pic high"><div class="chart"><div class="cover-fill" style="width: 85%;"></div><div class="cover-empty" style="width:15%;"></div></div></td>
	<td data-value="85" class="pct high">85%</td>
	<td data-value="20" class="abs high">17/20</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="84.21" class="pct high">84.21%</td>
	<td data-value="19" class="abs high">16/19</td>
	</tr>

<tr>
	<td class="file high" data-value="graphql.js"><a href="graphql.js.html">graphql.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="13" class="abs high">12/13</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	</tr>

<tr>
	<td class="file high" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="81.25" class="pic high"><div class="chart"><div class="cover-fill" style="width: 81%;"></div><div class="cover-empty" style="width:19%;"></div></div></td>
	<td data-value="81.25" class="pct high">81.25%</td>
	<td data-value="16" class="abs high">13/16</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="70" class="pct medium">70%</td>
	<td data-value="10" class="abs medium">7/10</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	</tr>

<tr>
	<td class="file medium" data-value="parsing.js"><a href="parsing.js.html">parsing.js</a></td>
	<td data-value="72.73" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 72%;"></div><div class="cover-empty" style="width:28%;"></div></div></td>
	<td data-value="72.73" class="pct medium">72.73%</td>
	<td data-value="11" class="abs medium">8/11</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="8" class="abs medium">6/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="72.73" class="pct medium">72.73%</td>
	<td data-value="11" class="abs medium">8/11</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Nov 12 2018 20:15:34 GMT-0800 (Pacific Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
