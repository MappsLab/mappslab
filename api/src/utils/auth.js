// @flow
import crypto from 'crypto-promise'
import jwt from 'jsonwebtoken'
import type { JWT, UserType } from 'Types/UserTypes'
import config from '../config'

const JWT_KEY = config.get('jwtKey')

export const createToken = (): Promise<string> =>
	new Promise((resolve, reject) => {
		crypto
			.randomBytes(48)
			.then((buffer) => {
				const token = buffer.toString('hex')
				resolve(token)
			})
			.catch((cryptErr) => {
				reject(cryptErr)
			})
	})

export const createJWT = (user: UserType): JWT => {
	const expiresIn = 1 * 24 * 60 * 60 // 1 day
	const { name, uid, roles } = user
	if (!name || !uid || !roles || !roles.length)
		throw new Error(`createJWT requires a name, uid, and roles`)
	const token = jwt.sign({ name, uid, roles }, JWT_KEY || 'abc', { expiresIn })
	const expires = new Date(Date.now() + expiresIn * 1000).toISOString() // send an actual date
	return {
		token: `Bearer ${token}`,
		expires,
	}
}

export const verifyJWT = (token: string): Promise<UserType> =>
	new Promise((resolve, reject) => {
		jwt.verify(token, JWT_KEY, (err, decoded) => {
			if (err) {
				reject(err)
				return
			}
			resolve(decoded)
		})
	})
