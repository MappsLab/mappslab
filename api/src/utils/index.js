// @flow
import { pipeP, flatten, filter, assoc, prop } from 'ramda'
import type { NestedArray } from 'ramda'

export const filterNullAndUndefined = (obj: Object): Promise<Object> =>
	new Promise((resolve) =>
		resolve(filter((a) => a !== null && a !== undefined)(obj)),
	)

// Makes regular functions async for promisePipe
export const promisify = (f: Function): Function => async (...args) =>
	f(...args)

export const promisePipe = (...funcs: Array<Function>) =>
	pipeP(...funcs.map(promisify))

export const arrayify = (...things: NestedArray<mixed>): Array<mixed> =>
	flatten(things)

export const slugify = (text: string) =>
	text
		.toString()
		.toLowerCase()
		.replace(/\s+/g, '-')
		.replace(/[^\w-]+/g, '')
		.replace(/--+/g, '-')
		.replace(/^-+/, '')
		.replace(/-+$/, '')

export const createSlugFrom = (key: string) => (obj: Object) =>
	assoc('slug', slugify(prop(key, obj)))(obj)
