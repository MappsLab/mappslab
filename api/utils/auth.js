// @flow
import crypto from 'crypto-promise'
import jwt from 'jsonwebtoken'
import type { JWT, UserType } from 'Types/UserTypes'
import { JWT_KEY } from '../config'

export const createToken = (): Promise<string> =>
	new Promise((resolve, reject) => {
		crypto
			.randomBytes(48)
			.then((buffer) => {
				const token = buffer.toString('hex')
				resolve(token)
			})
			.catch((cryptErr) => {
				reject(cryptErr)
			})
	})

export const createJWT = (user: UserType): JWT => {
	const expires = 1 * 24 * 60 * 60 // 1 day
	const { username, uid } = user
	const token = jwt.sign({ username, uid }, JWT_KEY || 'abc', { expiresIn: 10800 })
	return {
		token: `Bearer ${token}`,
		expires,
	}
}

export const verifyJWT = (token: string): Promise<UserType | Error> =>
	new Promise((resolve, reject) => {
		jwt.verify(token, JWT_KEY, (err, decoded) => {
			if (err) {
				reject(err)
				return
			}
			resolve(decoded)
		})
	})
