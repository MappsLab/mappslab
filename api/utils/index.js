// @flow
import * as R from 'ramda'

export const filterNullAndUndefined = (obj: Object): Promise<Object> =>
	new Promise((resolve) => resolve(R.filter((a) => a !== null && a !== undefined)(obj)))

// Makes regular functions async for promisePipe
export const promisify = (f: Function): Function => async (...args) => f(...args)

export const promisePipe = (...funcs: Array<Function>) => R.pipeP(...funcs.map(promisify))
