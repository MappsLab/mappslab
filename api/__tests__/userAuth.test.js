/* eslint-disable no-undef */
import { createJWT, verifyJWT } from 'Utils/auth'
import { request } from './utils/db'
import { getDBUsers } from './utils/user'
// import { joseph, teacher } from '../../../database/stubs/users'

/**
 * NOTE:
 * Only 'joseph' is seeded with a password: 'Password#1'
 * all other users have a temporaryPassword of `temporary`
 */
let users
let joseph
let teacher
let student

beforeAll(async (done) => {
	users = await getDBUsers()
	joseph = users.find((u) => u.email === 'joseph@good-idea.studio')
	teacher = users.find((u) => u.roles.includes('teacher'))
	student = users.find((u) => u.roles.includes('student'))
	done()
})

const uidLogin = /* GraphQL */ `
	mutation LoginViewer($password: String!, $uid: String, $email: String) {
		loginViewer(input: { uid: $uid, password: $password, email: $email }) {
			... on LoginSuccess {
				jwt {
					token
					expires
				}
				viewer {
					uid
					name
				}
			}
			... on RequiresReset {
				resetToken
			}
		}
	}
`

describe('queries', () => {
	it('[loginViewer] should return a jwt and viewer', async () => {
		const variables = { email: joseph.email, password: 'Password#1' }
		const result = await request(uidLogin, { variables })
		const { jwt, viewer } = result.data.loginViewer
		expect(/^Bearer/.test(jwt.token)).toBe(true)
		expect(jwt.expires).toBeGreaterThan(1)
		expect(viewer.name).toBe(joseph.name)
	})

	it('[loginViewer] should work with uid', async () => {
		const variables = { uid: joseph.uid, password: 'Password#1' }
		const result = await request(uidLogin, { variables })
		const { jwt, viewer } = result.data.loginViewer
		expect(/^Bearer/.test(jwt.token)).toBe(true)
		expect(jwt.expires).toBeGreaterThan(1)
		expect(viewer.name).toBe(joseph.name)
	})

	it('[loginViewer] should return `requiresReset` if the supplied password fails but matches `user.temporaryPassword`', async () => {
		const variables = { uid: student.uid, password: 'temporary' }
		const result = await request(uidLogin, { variables })
		expect(result.data.loginViewer.resetToken.length).toBe(96)
	})

	it('[loginViewer] should return a validation error with invalid credentials', async () => {
		const variables = { email: joseph.email, password: 'wrongPassword' }
		const result = await request(uidLogin, { variables })
		expect(result.errors[0].message).toBe('Email and password do not match')
	})

	it('[loginViewer] should return a validation error if user does not exist', async () => {
		const uidQuery = /* GraphQL */ `
			mutation LoginViewer($password: String!, $email: String!) {
				loginViewer(input: { email: $email, password: $password }) {
					... on LoginSuccess {
						jwt {
							token
							expires
						}
					}
				}
			}
		`
		const variables = { email: 'nobody@nowhere.com', password: 'password' }
		const result = await request(uidQuery, { variables })
		expect(result.errors[0].message).toBe('Email and password do not match')
	})

	it('[viewer] should return full user data given a JWT-parsed user in the context', async () => {
		const currentViewerQuery = /* GraphQL */ `
			{
				currentViewer {
					jwt {
						token
						expires
					}
					viewer {
						uid
						name
						email
					}
				}
			}
		`
		const originalViewer = createJWT(teacher)
		const JWTviewer = await verifyJWT(originalViewer.token.replace(/^Bearer /, ''))
		const context = { viewer: JWTviewer }
		const result = await request(currentViewerQuery, { context })
		const { viewer, jwt } = result.data.currentViewer
		expect(viewer.uid).toBe(teacher.uid)
		expect(viewer.email).toBe(teacher.email)
		expect(jwt.token).toMatch(/^Bearer /)
		expect(jwt.expires).toBe(86400)
	})

	describe('JWT creation & verification', async () => {
		it('should only return username, email, uid in JWT token', async () => {
			const jwt = createJWT(teacher)
			const result = await verifyJWT(jwt.token.replace(/^Bearer /, ''))
			expect(result).toHaveProperty('uid')
			expect(result).toHaveProperty('exp')
			expect(result).toHaveProperty('iat')
			expect(result).not.toHaveProperty('name')
			expect(result).not.toHaveProperty('disabled')
		})
	})
})

describe('[updatePassword]', () => {
	const getResetToken = async () => {
		const variables = { uid: student.uid, password: 'temporary' }
		const result = await request(uidLogin, { variables })
		return result.data.loginViewer.resetToken
	}

	const updatePasswordWithToken = /* GraphQL */ `
		mutation UpdatePassword($resetToken: String!, $password: String!) {
			updatePassword(input: { resetToken: $resetToken, password: $password }) {
				... on LoginSuccess {
					jwt {
						token
						expires
					}
					viewer {
						uid
						name
					}
				}
				... on RequiresReset {
					resetToken
				}
			}
		}
	`

	it('should throw an error for an invalid token', async () => {
		const variables = { password: 'newPassword', resetToken: 'invalid' }
		const result = await request(updatePasswordWithToken, { variables })
		expect(result.errors[0].message).toBe('This reset token is invalid')
	})

	it.skip('should throw an error for an expired token', async () => {
		/* Arrange */
		// const { container, getByTestId } = render( ... )
		/* Act */
		/* Assert */
		// expect(...)
	})

	it.skip('should reset a users password given a valid token', async () => {
		const resetToken = await getRestToken()
		const variables = { password: 'newPassword', resetToken }
		const result = await request(updatePasswordWithToken, { variables })
		expect(result.errors[0].message).toBe('This reset token is invalid')

		/* Arrange */
		// const { container, getByTestId } = render( ... )
		/* Act */
		/* Assert */
		// expect(...)
	})
})
