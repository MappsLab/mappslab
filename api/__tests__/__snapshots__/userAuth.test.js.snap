// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[resetPassword] should throw an error for an invalid token 1`] = `"This password reset request has expired."`;

exports[`[setTemporaryPassword] should disallow anyone from setting temporary passwords for admins 1`] = `
Array [
  [GraphQLError: Admins cannot have temporary passwords],
]
`;

exports[`[setTemporaryPassword] should disallow students from setting temporary passwords 1`] = `undefined`;
